share/ipld/nmt_hasher_integration_test.go

```diff
@@ -0,0 +1,162 @@
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	mrand "math/rand"
+
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	ds "github.com/ipfs/go-datastore"
+	dssync "github.com/ipfs/go-datastore/sync"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	"github.com/stretchr/testify/require"
+
+	"github.com/celestiaorg/celestia-node/share/availability/full"
+	availability_test "github.com/celestiaorg/celestia-node/share/availability/test"
+	"github.com/celestiaorg/celestia-node/share/service"
+)
+
+var _ blockstore.Blockstore = (*MockBlockstore)(nil)
+
+// CorruptBlock is a block where the cid doesn't match the data.
+type CorruptBlock struct {
+	cid  cid.Cid
+	data []byte
+}
+
+func (b *CorruptBlock) RawData() []byte {
+	return b.data
+}
+
+func (b *CorruptBlock) Cid() cid.Cid {
+	return b.cid
+}
+
+func (b *CorruptBlock) String() string {
+	return fmt.Sprintf("[Block %s]", b.Cid())
+}
+
+func (b *CorruptBlock) Loggable() map[string]interface{} {
+	return map[string]interface{}{
+		"block": b.Cid().String(),
+	}
+}
+
+func NewCorruptBlock(data []byte, fakeCID cid.Cid) *CorruptBlock {
+	return &CorruptBlock{
+		fakeCID,
+		data,
+	}
+}
+
+type MockBlockstore struct {
+	ds.Datastore
+	attacking bool
+}
+
+func (m MockBlockstore) Has(ctx context.Context, cid cid.Cid) (bool, error) {
+	return false, nil
+}
+
+func (m MockBlockstore) Get(ctx context.Context, cid cid.Cid) (blocks.Block, error) {
+	key := cid.String()
+	if m.attacking {
+		key = "corrupt" + key
+	}
+
+	data, err := m.Datastore.Get(ctx, ds.NewKey(key))
+	if err != nil {
+		return nil, err
+	}
+	return NewCorruptBlock(data, cid), nil
+}
+
+func (m MockBlockstore) GetSize(ctx context.Context, cid cid.Cid) (int, error) {
+	key := cid.String()
+	if m.attacking {
+		key = "corrupt" + key
+	}
+
+	return m.Datastore.GetSize(ctx, ds.NewKey(key))
+}
+
+func (m MockBlockstore) Put(ctx context.Context, block blocks.Block) error {
+	err := m.Datastore.Put(ctx, ds.NewKey(block.Cid().String()), block.RawData())
+	if err != nil {
+		return err
+	}
+
+	// create data that doesn't match the CID with arbitrary lengths between 0 and len(block.RawData())*2
+	corrupted := make([]byte, mrand.Int()%(len(block.RawData())*2))
+	mrand.Read(corrupted)
+	return m.Datastore.Put(ctx, ds.NewKey("corrupt"+block.Cid().String()), corrupted)
+}
+
+func (m MockBlockstore) PutMany(ctx context.Context, blocks []blocks.Block) error {
+	for _, b := range blocks {
+		err := m.Put(ctx, b)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m MockBlockstore) DeleteBlock(ctx context.Context, cid cid.Cid) error {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m MockBlockstore) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m MockBlockstore) HashOnRead(enabled bool) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func MockNode(t *testing.T, net *availability_test.DagNet) (*availability_test.Node, *MockBlockstore) {
```

Comment:
does this need to be exported? If not, I'd recommend not exporting test methods by default.

Reply:
It doesn't need to be, I will unexport it. @Wondertan How does exporting from _test.go work? Does it actually get exported + end up in the binary?

Reply:
`_test.go` can only export from each other. You cannot import anything defined in a test file from a non-test file. Same as you cannot import anything from the `main` pkg.

I do agree, though, that in this case is better to keep the test utility lowercased for it to be perceived as unexported

Reply:
> _test.go can only export from each other. You cannot import anything defined in a test file from a non-test file.

tbh TIL haha has this always been the case? Or is that a recent version update?

Maybe I've just been confusing exported methods in a testing pkg and exported methods in `_test.go` files.  But good to know

Reply:
> tbh TIL haha has this always been the case? Or is that a recent version update?

I remember it has always been like this, but I am in Golang for only 3-4 years in this change could land before.

Reply:
These are now located in the correct spot, but are exported so that the actual `_test.go` can access them. I renamed the struct types so that it is clear that everything is a `TestNode` on a `TestDagNet`, and not some actual type. Regardless, the constructor takes a `testing.T`, so it can only be used by tests.